jetty-io/src/main/java/org/eclipse/jetty/io/BufferCallbackAccumulator.java

```diff
@@ -0,0 +1,93 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+
+public class BufferCallbackAccumulator
+{
+    private final List<Entry> _entries = new ArrayList<>();
+
+    private static class Entry
+    {
+        private final ByteBuffer buffer;
+        private final Callback callback;
+
+        Entry(ByteBuffer buffer, Callback callback)
+        {
+            this.buffer = buffer;
+            this.callback = callback;
+        }
+    }
+
+    public void addEntry(ByteBuffer buffer, Callback callback)
+    {
+        _entries.add(new Entry(buffer, callback));
+    }
+
+    /**
+     * Get the amount of bytes which have been accumulated.
+     * This will add up the remaining of each buffer in the accumulator.
+     * @return the total length of the content in the accumulator.
+     */
+    public int getLength()
+    {
+        int length = 0;
+        for (Entry entry : _entries)
+            length = Math.addExact(length, entry.buffer.remaining());
```

Comment:
Surely there is some unintelligble stream based way of doing this:
```suggestion
        length = _entries.stream().map(Entry::getBuffer).map(ByteBuffer::remaining).sum();
```

Reply:
I'll replace this with a counter as they are added.